src/Monkeypatcher.cc

```diff
@@ -428,15 +434,36 @@ bool Monkeypatcher::try_patch_syscall(RecordTask* t) {
     return false;
   }
 
-  uint8_t following_bytes[256];
+  static const intptr_t MAXIMUM_LOOKBACK = 6;
+  uint8_t bytes[256 + MAXIMUM_LOOKBACK];
   size_t bytes_count = t->read_bytes_fallible(
-      ip.to_data_ptr<uint8_t>(), sizeof(following_bytes), following_bytes);
+      ip.to_data_ptr<uint8_t>() - MAXIMUM_LOOKBACK, sizeof(bytes), bytes);
+  if (bytes_count < MAXIMUM_LOOKBACK) {
+    LOG(debug) << "Declining to patch syscall at " << ip << " for lack of lookback";
+    tried_to_patch_syscall_addresses.insert(ip);
+    return false;
+  }
+  size_t following_bytes_count = bytes_count - MAXIMUM_LOOKBACK;
+  uint8_t* following_bytes = &bytes[MAXIMUM_LOOKBACK];
 
   intptr_t syscallno = r.original_syscallno();
   for (auto& hook : syscall_hooks) {
-    if (bytes_count >= hook.next_instruction_length &&
-        memcmp(following_bytes, hook.next_instruction_bytes,
-               hook.next_instruction_length) == 0) {
+    if ((!(hook.flags & PATCH_SYSCALL_INSTRUCTION_IS_LAST) &&
+         following_bytes_count >= hook.next_instruction_length &&
+         memcmp(following_bytes, hook.next_instruction_bytes,
+                hook.next_instruction_length) == 0) ||
+        (!!(hook.flags & PATCH_SYSCALL_INSTRUCTION_IS_LAST) &&
+         bytes_count >= hook.next_instruction_length + (size_t)rr::syscall_instruction_length(arch) &&
+         memcmp(bytes, hook.next_instruction_bytes,
+                hook.next_instruction_length) == 0)) {
```

Comment:
I think this would be easier to read with a boolean variable `matches_hook` and an `if (hook.flags & PATCH_SYSCALL_INSTRUCTION_IS_LAST)` if/else block that sets that variable.